core/services/relay/evm/statuschecker/txm_status_checker.go

```diff
@@ -0,0 +1,66 @@
+package statuschecker
+
+import (
+	"context"
+	"fmt"
+	"strings"
+)
+
+//go:generate mockery --quiet --name TransactionStatusChecker --output ../mocks/ --case=underscore
+
+// TODO replace with actual implementation coming from "github.com/smartcontractkit/chainlink-common/pkg/types"
+type TransactionStatus int
+
+const (
+	Unknown TransactionStatus = iota
+	Unconfirmed
+	Finalized
+	Failed
+	Fatal
+)
+
+type TxManager interface {
+	GetTransactionStatus(ctx context.Context, transactionID string) (TransactionStatus, error)
+}
+
+type TransactionStatusChecker interface {
+	CheckMessageStatus(ctx context.Context, msgID string) ([]TransactionStatus, int, error)
+}
+
+type TxmStatusChecker struct {
+	txManager TxManager
+}
+
+type NoOpTxManager struct{}
+
+func (n *NoOpTxManager) GetTransactionStatus() error {
+	return nil
+}
+
+func NewTransactionStatusChecker(txManager TxManager) *TxmStatusChecker {
+	return &TxmStatusChecker{txManager: txManager}
+}
+
+// CheckMessageStatus checks the status of a message by checking the status of all transactions associated with the message ID.
+// It returns a slice of all statuses and the number of transactions found (-1 if none).
+// The key will follow the format: <msgID>-<counter>. TXM will be queried for each key until a NotFound error is returned.
+// The goal is to find all transactions associated with a message ID and snooze messages if they are fatal in the Execution Plugin.
+func (tsc *TxmStatusChecker) CheckMessageStatus(ctx context.Context, msgID string) ([]TransactionStatus, int, error) {
+	var allStatuses []TransactionStatus
+	var counter int
+
+	for {
+		transactionID := fmt.Sprintf("%s-%d", msgID, counter)
+		status, err := tsc.txManager.GetTransactionStatus(ctx, transactionID)
+		if err != nil {
+			if status == Unknown && strings.Contains(err.Error(), fmt.Sprintf("failed to find transaction with IdempotencyKey %s", transactionID)) {
```

Comment:
https://github.com/smartcontractkit/chainlink/pull/13040/files

```
func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) GetTransactionStatus(ctx context.Context, transactionID string) (status commontypes.TransactionStatus, err error) {
	// Loads attempts and receipts in the transaction
	tx, err := b.txStore.FindTxWithIdempotencyKey(ctx, transactionID, b.chainID)
	if err != nil {
		return status, fmt.Errorf("failed to find transaction with IdempotencyKey %s: %w", transactionID, err)
	}
	// This check is required since a no-rows error returns nil err
	if tx == nil {
		return status, fmt.Errorf("failed to find transaction with IdempotencyKey %s", transactionID)
	}
	switch tx.State {
	case TxUnconfirmed, TxConfirmedMissingReceipt:
		// Return unconfirmed for ConfirmedMissingReceipt since a receipt is required to determine if it is finalized
		return commontypes.Unconfirmed, nil
	case TxConfirmed:
		// TODO: Check for finality and return finalized status
		// Return unconfirmed if tx receipt's block is newer than the latest finalized block
		return commontypes.Unconfirmed, nil
	case TxFatalError:
		// Use an ErrorClassifier to determine if the transaction is considered Fatal
		txErr := b.newErrorClassifier(tx.GetError())
		if txErr != nil && txErr.IsFatal() {
			return commontypes.Fatal, tx.GetError()
		}
		// Return failed for all other tx's marked as FatalError
		return commontypes.Failed, tx.GetError()
	default:
		// Unstarted and InProgress are classified as unknown since they are not supported by the ChainWriter interface
		return commontypes.Unknown, nil
	}
}
```

Reply:
Status will always return the default value `Unknown` in this case. But I agree that this check is useless. We could just rely on the error check.